generator/internal/rust/templates/common/enum.mustache

```diff
@@ -18,60 +18,128 @@ limitations under the License.
 {{{.}}}
 {{/Codec.DocLines}}
 {{> /templates/common/feature_gate}}
-#[derive(Clone, Debug, PartialEq, serde::Deserialize, serde::Serialize)]
-pub struct {{Codec.Name}}(i32);
-
-{{> /templates/common/feature_gate}}
-impl {{Codec.Name}} {
+#[derive(Clone, Debug, PartialEq)]
+#[non_exhaustive]
+pub enum {{Codec.Name}} {
     {{#Codec.UniqueNames}}
-
     {{#Codec.DocLines}}
     {{{.}}}
     {{/Codec.DocLines}}
-    pub const {{Codec.Name}}: {{Codec.EnumType}} = {{Codec.EnumType}}::new({{Number}});
+    {{Codec.VariantName}},
     {{/Codec.UniqueNames}}
+    /// If set, the enum was initialized with an unknown value.
+    ///
+    /// Applications can examine the value using [{{Codec.Name}}::value] or
+    /// [{{Codec.Name}}::name].
+    UnknownValue({{Codec.ModuleName}}::UnknownValue),
+}
 
-    /// Creates a new {{Codec.Name}} instance.
-    pub(crate) const fn new(value: i32) -> Self {
-        Self(value)
-    }
+{{!
+    The `UnknownValue` variant must be public, but we do not want people to
+    build that variant with a known value. We use a struct, with a `pub(crate)`
+    constructor, and a `pub(crate)` enum.
+}}
+#[doc(hidden)]
+{{> /templates/common/feature_gate}}
+pub mod {{Codec.ModuleName}} {
+    {{! Very rarely, this is unused. It is easier to always disable the warning. }}
+    #[allow(unused_imports)]
+    use super::*;
```

Comment:
I don't see us using any types from `super::`

Reply:
Depending on how the code is generated then `wkt::` may be a top-level external crate, may be a module in super, or may be an external crate imported via `extern crate` and carried to each module via these `use super::*`.

So, using `wkt::` below works most times (but not when it is a module), `::wkt` works other times, and `super::wkt` works in yet other times.  The only approach that works always in what we have here.

We may want to clean this up, but I think the biggest obstable are the gRPC clients.

Reply:
Got it, thanks. I forgot about the generated types in `wkt` (despite looking at them last night)